# Energy Supply and Renewable Electricity Production Project
------------------------------------------------------------------------------
## Implemented Process in this project:
- Importing my packages [`pandas` and `numpy`].

- Importing my datasets [`3 datasets`].

- Discovering and exploring my data through doing some EDA by utilizing pandas functions.

- Cleanign data by replacing some unvalid data/dropping some and handelling the rest.

- Joining data using pandas.

- Preparing and filterting `merged_df` dataframe for getting some insights from it.

- Getting some statistical insights by using some of pandas and numpy functions.
-----------------------------------------------------------------------------------------------
## Objectives:
- Data Cleaning
- Data Transformation
---------------------------------------------------------------------------------------------------
## Questions that are answered in this project:
- What is the Average `GDP` over the last 10 years for each country?

- What is the mean of `Energy Supply per Capita`?

- What country has the maximum `% Reneweble`? And what is the percentage?

- What are the ratios of `Self-Citations` to `Total-Citations`? And what is the maximum value?

- Which country has the maximum `ratio`?

- What is the distribution of my `% Reneweble` column? (through knowing how many values are above or below the median)

- What is the `size`, `Sum`, `mean`and the `STD` for the `estimated population` for each country? 

## GETTING STARTED:

### STEP 0
Loading the energy dataset which is a list of indicators of energy supply and renewable electricity production from the United Nations for the year 2013.


In [1]:
# importing packages [pandas , numpy]
import pandas as pd 
import numpy as np

In [2]:
# reading the dataset which is in excel format, with excluding the header and the footer
energy = pd.read_excel('Energy Indicators.xls' , header = 17 , skipfooter = 38)

In [63]:
# checking my dataframe
energy

Country  Energy Supply  Energy Supply per Capita  % Reneweble
0       Afghanistan   3.210000e+08                      10.0    78.669280
1           Albania   1.020000e+08                      35.0   100.000000
2           Algeria   1.959000e+09                      51.0     0.551010
3    American Samoa            NaN                       NaN     0.641026
4           Andorra   9.000000e+06                     121.0    88.695650
..              ...            ...                       ...          ...
222        Viet Nam   2.554000e+09                      28.0    45.321520
223     Wallis and    0.000000e+00                      26.0     0.000000
224           Yemen   3.440000e+08                      13.0     0.000000
225          Zambia   4.000000e+08                      26.0    99.714670
226        Zimbabwe   4.800000e+08                      32.0    52.536120

[227 rows x 4 columns]

----------------------------------------------------------------------------
Dropping first two column of the dataframe

In [4]:
# dropping the first two columns ('Unnamed: 0' , 'Unnamed: 1')
energy = energy.drop(['Unnamed: 0', 'Unnamed: 1'] , axis = 1)

In [5]:
energy

Unnamed: 2 Petajoules Gigajoules           %
0                  Afghanistan        321         10   78.669280
1                      Albania        102         35  100.000000
2                      Algeria       1959         51    0.551010
3               American Samoa        ...        ...    0.641026
4                      Andorra          9        121   88.695650
..                         ...        ...        ...         ...
222                   Viet Nam       2554         28   45.321520
223  Wallis and Futuna Islands          0         26    0.000000
224                      Yemen        344         13    0.000000
225                     Zambia        400         26   99.714670
226                   Zimbabwe        480         32   52.536120

[227 rows x 4 columns]

In [6]:
# renaming the other 4 columns with much more descriptive names
energy.rename(columns = {'Unnamed: 2' : 'Country' ,
                        'Petajoules':'Energy Supply',
                        'Gigajoules' : 'Energy Supply per Capita',
                        '%' : '% Reneweble'} , inplace = True)

In [7]:
# checking my data with new column's names
energy

Country Energy Supply Energy Supply per Capita  \
0                  Afghanistan           321                       10   
1                      Albania           102                       35   
2                      Algeria          1959                       51   
3               American Samoa           ...                      ...   
4                      Andorra             9                      121   
..                         ...           ...                      ...   
222                   Viet Nam          2554                       28   
223  Wallis and Futuna Islands             0                       26   
224                      Yemen           344                       13   
225                     Zambia           400                       26   
226                   Zimbabwe           480                       32   

     % Reneweble  
0      78.669280  
1     100.000000  
2       0.551010  
3       0.641026  
4      88.695650  
..           ...  
222    45.321520  
223     0.000000  
224     0.000000  
225    99.714670  
226    52.536120  

[227 rows x 4 columns]

In [8]:
# replacing any missing values with np.nan 
energy.replace(to_replace= '...' , value = np.nan , inplace = True)

In [62]:
energy

Country  Energy Supply  Energy Supply per Capita  % Reneweble
0       Afghanistan   3.210000e+08                      10.0    78.669280
1           Albania   1.020000e+08                      35.0   100.000000
2           Algeria   1.959000e+09                      51.0     0.551010
3    American Samoa            NaN                       NaN     0.641026
4           Andorra   9.000000e+06                     121.0    88.695650
..              ...            ...                       ...          ...
222        Viet Nam   2.554000e+09                      28.0    45.321520
223     Wallis and    0.000000e+00                      26.0     0.000000
224           Yemen   3.440000e+08                      13.0     0.000000
225          Zambia   4.000000e+08                      26.0    99.714670
226        Zimbabwe   4.800000e+08                      32.0    52.536120

[227 rows x 4 columns]

------------------------------------------------------------------------------------------------------------------------------
### Converting each value in 'Energy Supply' from a petajoules to gigajoules

In [10]:
# converting each value in 'Energy Supply' to gigajoules
energy['Energy Supply']=energy['Energy Supply'].apply(lambda x : x*1000000)

In [11]:
energy

Country  Energy Supply  Energy Supply per Capita  \
0                  Afghanistan   3.210000e+08                      10.0   
1                      Albania   1.020000e+08                      35.0   
2                      Algeria   1.959000e+09                      51.0   
3               American Samoa            NaN                       NaN   
4                      Andorra   9.000000e+06                     121.0   
..                         ...            ...                       ...   
222                   Viet Nam   2.554000e+09                      28.0   
223  Wallis and Futuna Islands   0.000000e+00                      26.0   
224                      Yemen   3.440000e+08                      13.0   
225                     Zambia   4.000000e+08                      26.0   
226                   Zimbabwe   4.800000e+08                      32.0   

     % Reneweble  
0      78.669280  
1     100.000000  
2       0.551010  
3       0.641026  
4      88.695650  
..           ...  
222    45.321520  
223     0.000000  
224     0.000000  
225    99.714670  
226    52.536120  

[227 rows x 4 columns]

In [12]:
# replacing the list of countries in 'Country' column
energy.replace({'Country' : {"Republic of Korea": "South Korea" ,
                            "United States of America": "United States",
                            "United Kingdom of Great Britain and Northern Ireland": "United Kingdom",
                            "China, Hong Kong Special Administrative Region": "Hong Kong"}} , inplace = True)

In [13]:
# validing the relplacing function
energy[energy['Country'] == 'South Korea']

Country  Energy Supply  Energy Supply per Capita  % Reneweble
164  South Korea   1.100700e+10                     221.0     2.279353

In [14]:
# checking on countries' values that have numbers and/or parenthesis
energy[energy['Country'] == 'Bolivia (Plurinational State of)']

Country  Energy Supply  Energy Supply per Capita  \
24  Bolivia (Plurinational State of)    336000000.0                      32.0   

    % Reneweble  
24     31.47712

In [15]:
# checking on contries' values that have numbers and/or parenthesis
energy[energy['Country'] == 'Switzerland17']

Country  Energy Supply  Energy Supply per Capita  % Reneweble
197  Switzerland17   1.113000e+09                     136.0     57.74548

In [16]:
# Remove any value in a Country column that has number or parenthesis [al]
energy['Country'] = energy['Country'].str.extract('(\w+\s\w+[^0-9]|\w+[^0-9])')

In [17]:
# checking the first 30 rows to check row number 24 that was having 'Bolivia (Plurinational State of)' value with parenthesis
energy['Country'].head(30)

0         Afghanistan
1             Albania
2             Algeria
3      American Samoa
4             Andorra
5              Angola
6            Anguilla
7        Antigua and 
8           Argentina
9             Armenia
10              Aruba
11          Australia
12            Austria
13         Azerbaijan
14            Bahamas
15            Bahrain
16         Bangladesh
17           Barbados
18            Belarus
19            Belgium
20             Belize
21              Benin
22            Bermuda
23             Bhutan
24           Bolivia 
25           Bonaire,
26        Bosnia and 
27           Botswana
28             Brazil
29    British Virgin 
Name: Country, dtype: object

In [18]:
# validating my code by checking on a value with a number that was existing in a row number 197
energy[energy['Country'] == 'Switzerland17']

Empty DataFrame
Columns: [Country, Energy Supply, Energy Supply per Capita, % Reneweble]
Index: []

----------------------------------------------------------------------------------
## STEP 1
Next step, loading the second dataset, the GDP data, which is a csv containing countries' GDP from 1960 to 2015 from World Bank. 

reading dataset with skipping header

In [19]:
# loading the GDP data from the file [second dataset]
GDP = pd.read_csv('world_bank.csv' , header = 4)
GDP

Country Name Country Code                            Indicator Name  \
0               Aruba          ABW  GDP at market prices (constant 2010 US$)   
1             Andorra          AND  GDP at market prices (constant 2010 US$)   
2         Afghanistan          AFG  GDP at market prices (constant 2010 US$)   
3              Angola          AGO  GDP at market prices (constant 2010 US$)   
4             Albania          ALB  GDP at market prices (constant 2010 US$)   
..                ...          ...                                       ...   
259       Yemen, Rep.          YEM  GDP at market prices (constant 2010 US$)   
260      South Africa          ZAF  GDP at market prices (constant 2010 US$)   
261  Congo, Dem. Rep.          COD  GDP at market prices (constant 2010 US$)   
262            Zambia          ZMB  GDP at market prices (constant 2010 US$)   
263          Zimbabwe          ZWE  GDP at market prices (constant 2010 US$)   

     Indicator Code          1960          1961          1962          1963  \
0    NY.GDP.MKTP.KD           NaN           NaN           NaN           NaN   
1    NY.GDP.MKTP.KD           NaN           NaN           NaN           NaN   
2    NY.GDP.MKTP.KD           NaN           NaN           NaN           NaN   
3    NY.GDP.MKTP.KD           NaN           NaN           NaN           NaN   
4    NY.GDP.MKTP.KD           NaN           NaN           NaN           NaN   
..              ...           ...           ...           ...           ...   
259  NY.GDP.MKTP.KD           NaN           NaN           NaN           NaN   
260  NY.GDP.MKTP.KD  7.707081e+10  8.003449e+10  8.497844e+10  9.124438e+10   
261  NY.GDP.MKTP.KD  1.508024e+10  1.344383e+10  1.629401e+10  1.714358e+10   
262  NY.GDP.MKTP.KD  4.592975e+09  4.655503e+09  4.539542e+09  4.688093e+09   
263  NY.GDP.MKTP.KD  3.338344e+09  3.549199e+09  3.600111e+09  3.824915e+09   

             1964          1965  ...          2006          2007  \
0             NaN           NaN  ...           NaN           NaN   
1             NaN           NaN  ...  4.018196e+09  4.021331e+09   
2             NaN           NaN  ...  1.030523e+10  1.172119e+10   
3             NaN           NaN  ...  5.581103e+10  6.842044e+10   
4             NaN           NaN  ...  9.771760e+09  1.034829e+10   
..            ...           ...  ...           ...           ...   
259           NaN           NaN  ...  2.672565e+10  2.761787e+10   
260  9.848904e+10  1.072507e+11  ...  3.402852e+11  3.585261e+11   
261  1.672524e+10  1.689205e+10  ...  1.650894e+10  1.754232e+10   
262  5.260699e+09  6.136472e+09  ...  1.440569e+10  1.560892e+10   
263  3.782605e+09  3.968352e+09  ...  1.006276e+10  9.695130e+09   

             2008          2009          2010          2011          2012  \
0             NaN           NaN  2.467704e+09           NaN           NaN   
1    3.675728e+09  3.535389e+09  3.346317e+09  3.185605e+09  3.129538e+09   
2    1.214448e+10  1.469733e+10  1.593680e+10  1.691113e+10  1.935220e+10   
3    7.787420e+10  7.975320e+10  8.247091e+10  8.570262e+10  9.012096e+10   
4    1.112752e+10  1.150029e+10  1.192695e+10  1.223109e+10  1.240477e+10   
..            ...           ...           ...           ...           ...   
259  2.872656e+10  2.991436e+10  3.090675e+10  2.624342e+10  2.689160e+10   
260  3.699668e+11  3.642764e+11  3.753494e+11  3.874074e+11  3.960071e+11   
261  1.863448e+10  1.916651e+10  2.052329e+10  2.193213e+10  2.350200e+10   
262  1.682234e+10  1.837342e+10  2.026555e+10  2.140358e+10  2.302438e+10   
263  7.982103e+09  8.459783e+09  9.422161e+09  1.054391e+10  1.165789e+10   

             2013          2014          2015  
0             NaN           NaN           NaN  
1    3.127550e+09           NaN           NaN  
2    1.973134e+10  1.999032e+10  2.029415e+10  
3    9.626143e+10  1.008863e+11  1.039106e+11  
4    1.254247e+10  1.279331e+10  1.312082e+10  
..            ...           ...           ...  
259  2.800914e+10    

In [20]:
# checking on the old values data before replaccing them
GDP[GDP['Country Name'] == 'Korea, Rep.']

Country Name Country Code                            Indicator Name  \
123  Korea, Rep.          KOR  GDP at market prices (constant 2010 US$)   

     Indicator Code          1960          1961          1962          1963  \
123  NY.GDP.MKTP.KD  2.757696e+10  2.893825e+10  2.964966e+10  3.247578e+10   

             1964          1965  ...          2006          2007  \
123  3.493136e+10  3.674434e+10  ...  9.410199e+11  9.924316e+11   

             2008          2009          2010          2011          2012  \
123  1.020510e+12  1.027730e+12  1.094499e+12  1.134796e+12  1.160809e+12   

             2013          2014          2015  
123  1.194429e+12  1.234340e+12  1.266580e+12  

[1 rows x 60 columns]

In [21]:
# replacing values with anothers in 'Country Name' column
GDP.replace({'Country Name' : {"Korea, Rep.": "South Korea",
                               "Iran, Islamic Rep.": "Iran",
                               "Hong Kong SAR, China": "Hong Kong"}} , inplace = True)

In [22]:
# validating the result
GDP[GDP['Country Name'] == 'Korea, Rep.']

Empty DataFrame
Columns: [Country Name, Country Code, Indicator Name, Indicator Code, 1960, 1961, 1962, 1963, 1964, 1965, 1966, 1967, 1968, 1969, 1970, 1971, 1972, 1973, 1974, 1975, 1976, 1977, 1978, 1979, 1980, 1981, 1982, 1983, 1984, 1985, 1986, 1987, 1988, 1989, 1990, 1991, 1992, 1993, 1994, 1995, 1996, 1997, 1998, 1999, 2000, 2001, 2002, 2003, 2004, 2005, 2006, 2007, 2008, 2009, 2010, 2011, 2012, 2013, 2014, 2015]
Index: []

[0 rows x 60 columns]

-----------------------------------------------------------------------------------------------------------------------
## STEP 2
Finally, load the third and the last Sciamgo Journal and Country Rank data for Energy Engineering and Power Technology dataset, 
which ranks countries based on their journal contributions in the aforementioned area.

In [23]:
# loading the third dataset
ScimEn = pd.read_excel('scimagojr-3.xlsx')
ScimEn

Rank             Country  Documents  Citable documents  Citations  \
0       1               China     127050             126767     597237   
1       2       United States      96661              94747     792274   
2       3               Japan      30504              30287     223024   
3       4      United Kingdom      20944              20357     206091   
4       5  Russian Federation      18534              18301      34266   
..    ...                 ...        ...                ...        ...   
186   187              Guyana          1                  1          0   
187   188    Christmas Island          1                  1          0   
188   189             Reunion          1                  1          2   
189   190         Saint Lucia          1                  1          0   
190   191          Mauritania          1                  1          1   

     Self-citations  Citations per document  H index  
0            411683                    4.70      138  
1            265436                    8.20      230  
2             61554                    7.31      134  
3             37874                    9.84      139  
4             12422                    1.85       57  
..              ...                     ...      ...  
186               0                    0.00        0  
187               0                    0.00        0  
188               1                    2.00        1  
189               0                    0.00        0  
190               0                    1.00        1  

[191 rows x 8 columns]

-----------------------------------------------------------------------------------------------------------------------------
## STEP 3
Join the three dataframes: GDP, Energy, and ScimEn into a new dataset (using the intersection of country names). Use only the last 10 years (2006-2015) of GDP data and only the top 15 countries by Scimagojr 'Rank' (Rank 1 through 15).


In [24]:
# rename the 'Country Name' column in GDP dataframe in order to be able to merge the 3 dfs based on their 'Country' values
GDP.rename(columns = {'Country Name' : 'Country'} , inplace = True)
GDP.head(3)

Country Country Code                            Indicator Name  \
0        Aruba          ABW  GDP at market prices (constant 2010 US$)   
1      Andorra          AND  GDP at market prices (constant 2010 US$)   
2  Afghanistan          AFG  GDP at market prices (constant 2010 US$)   

   Indicator Code  1960  1961  1962  1963  1964  1965  ...          2006  \
0  NY.GDP.MKTP.KD   NaN   NaN   NaN   NaN   NaN   NaN  ...           NaN   
1  NY.GDP.MKTP.KD   NaN   NaN   NaN   NaN   NaN   NaN  ...  4.018196e+09   
2  NY.GDP.MKTP.KD   NaN   NaN   NaN   NaN   NaN   NaN  ...  1.030523e+10   

           2007          2008          2009          2010          2011  \
0           NaN           NaN           NaN  2.467704e+09           NaN   
1  4.021331e+09  3.675728e+09  3.535389e+09  3.346317e+09  3.185605e+09   
2  1.172119e+10  1.214448e+10  1.469733e+10  1.593680e+10  1.691113e+10   

           2012          2013          2014          2015  
0           NaN           NaN           NaN           NaN  
1  3.129538e+09  3.127550e+09           NaN           NaN  
2  1.935220e+10  1.973134e+10  1.999032e+10  2.029415e+10  

[3 rows x 60 columns]

In [25]:
# create a new df that has the 3 data frame merged together
merged_df = pd.merge(ScimEn , pd.merge(energy, GDP ,how= 'left', on='Country'),how='left', on='Country')
merged_df 

Rank             Country  Documents  Citable documents  Citations  \
0       1               China     127050             126767     597237   
1       2       United States      96661              94747     792274   
2       3               Japan      30504              30287     223024   
3       4      United Kingdom      20944              20357     206091   
4       5  Russian Federation      18534              18301      34266   
..    ...                 ...        ...                ...        ...   
186   187              Guyana          1                  1          0   
187   188    Christmas Island          1                  1          0   
188   189             Reunion          1                  1          2   
189   190         Saint Lucia          1                  1          0   
190   191          Mauritania          1                  1          1   

     Self-citations  Citations per document  H index  Energy Supply  \
0            411683                    4.70      138   1.271910e+11   
1            265436                    8.20      230            NaN   
2             61554                    7.31      134   1.898400e+10   
3             37874                    9.84      139            NaN   
4             12422                    1.85       57   3.070900e+10   
..              ...                     ...      ...            ...   
186               0                    0.00        0   3.400000e+07   
187               0                    0.00        0            NaN   
188               1                    2.00        1            NaN   
189               0                    0.00        0   6.000000e+06   
190               0                    1.00        1   5.100000e+07   

     Energy Supply per Capita  ...          2006          2007          2008  \
0                        93.0  ...  3.992331e+12  4.559041e+12  4.997775e+12   
1                         NaN  ...           NaN           NaN           NaN   
2                       149.0  ...  5.496542e+12  5.617036e+12  5.558527e+12   
3                         NaN  ...           NaN           NaN           NaN   
4                       214.0  ...  1.385793e+12  1.504071e+12  1.583004e+12   
..                        ...  ...           ...           ...           ...   
186                      45.0  ...  1.919759e+09  2.054533e+09  2.095143e+09   
187                       NaN  ...           NaN           NaN           NaN   
188                       NaN  ...           NaN           NaN           NaN   
189                      33.0  ...           NaN           NaN           NaN   
190                      13.0  ...  4.025616e+09  4.139051e+09  4.183751e+09   

             2009          2010          2011          2012          2013  \
0    5.459247e+12  6.039659e+12  6.612490e+12  7.124978e+12  7.672448e+12   
1             NaN           NaN           NaN           NaN           NaN   
2    5.251308e+12  5.498718e+12  5.473738e+12  5.569102e+12  5.644659e+12   
3             NaN           NaN           NaN           NaN           NaN   
4    1.459199e+12  1.524917e+12  1.589943e+12  1.645876e+12  1.666934e+12   
..            ...           ...           ...           ...           ...   
186  2.164673e+09  2.259288e+09  2.382129e+09  2.496863e+09  2.627240e+09   
187           NaN           NaN           NaN           NaN           NaN   
188           NaN           NaN           NaN           NaN           NaN   
189           NaN           NaN           NaN           NaN           NaN   
190  4.140153e+09  4.337792e+09  4.541844e+09  4.805048e+09  5.097687e+09   

             2014          2015  
0    8.230121e+12  8.797999e+12  
1             NaN           NaN  
2    5.642884e+12  5.669563e+12  
3             NaN           NaN  
4    1.678709e+12  1.616149e+12  
..            ...           ...  
186  2.728150e+09  2.810153e+09  
187           NaN           NaN  
188           NaN           NaN  
189           NaN           NaN  
190

In [26]:
# subseting the new dataframe 'merged_df' by specific columns & top 15 countries with the highest ranking values
merged_df = merged_df.iloc[:15]
merged_df = merged_df.set_index('Country')
merged_df = merged_df[['Rank', 'Documents', 'Citable documents', 'Citations', 'Self-citations', 
                       'Citations per document', 'H index',
                       'Energy Supply', 'Energy Supply per Capita', '% Reneweble', '2006', '2007',
                       '2008', '2009', '2010', '2011', '2012', '2013', '2014', '2015']]
merged_df

Rank  Documents  Citable documents  Citations  \
Country                                                             
China                  1     127050             126767     597237   
United States          2      96661              94747     792274   
Japan                  3      30504              30287     223024   
United Kingdom         4      20944              20357     206091   
Russian Federation     5      18534              18301      34266   
Canada                 6      17899              17620     215003   
Germany                7      17027              16831     140566   
India                  8      15005              14841     128763   
France                 9      13153              12973     130632   
South Korea           10      11983              11923     114675   
Italy                 11      10964              10794     111850   
Spain                 12       9428               9330     123336   
Iran                  13       8896               8819      57470   
Australia             14       8831               8725      90765   
Brazil                15       8668               8596      60702   

                    Self-citations  Citations per document  H index  \
Country                                                               
China                       411683                    4.70      138   
United States               265436                    8.20      230   
Japan                        61554                    7.31      134   
United Kingdom               37874                    9.84      139   
Russian Federation           12422                    1.85       57   
Canada                       40930                   12.01      149   
Germany                      27426                    8.26      126   
India                        37209                    8.58      115   
France                       28601                    9.93      114   
South Korea                  22595                    9.57      104   
Italy                        26661                   10.20      106   
Spain                        23964                   13.08      115   
Iran                         19125                    6.46       72   
Australia                    15606                   10.28      107   
Brazil                       14396                    7.00       86   

                    Energy Supply  Energy Supply per Capita  % Reneweble  \
Country                                                                    
China                1.271910e+11                      93.0    19.754910   
United States                 NaN                       NaN          NaN   
Japan                1.898400e+10                     149.0    10.232820   
United Kingdom                NaN                       NaN          NaN   
Russian Federation   3.070900e+10                     214.0    17.288680   
Canada               1.043100e+10                     296.0    61.945430   
Germany              1.326100e+10                     165.0    17.901530   
India                3.319500e+10                      26.0    14.969080   
France               1.059700e+10                     166.0    17.020280   
South Korea          1.100700e+10                     221.0     2.279353   
Italy                6.530000e+09                     109.0    33.667230   
Spain                4.923000e+09                     106.0    37.968590   
Iran                          NaN                       NaN          NaN   
Australia            5.386000e+09                     231.0    11.810810   
Brazil               1.214900e+10                      59.0    69.648030   

                            2006          2007          2008          2009  \
Country                                                                      
China               3.992331e+12  4.559041e+12  4.997775e+12  5.459247e+12   
United States                NaN           NaN           NaN           NaN   
Japan               5.496542e+12 

In [27]:
# Another way of filtering the 'merged_df' based on rank value
# merged_df = merged_df.query(("`Rank` <= 15"))
# merged_df

In [28]:
# validating the shape of the new data frame which should be 15 x 20
merged_df.shape

(15, 20)

-------------------------------------------------------------------------------------------
## STEP 4 : How many rows have been lost in 'STEP 3' ?
In the previous code has joined three datasets, then reduced this 'merged_df' to just the top 15 entries. 
When we joined the datasets, but before reducing this to the top 15 items, how many entries have we loset?

This step should yield a single number.

## SOL :
Before i filter based on the Rank column level, the 'merged_df' had 191 row,,
then i filtered  by usinf iloc[] function to get the top 15 countries 

So, The number of lost entries = 191 - 15 = 176 row

In [29]:
# validating result:
# creating a new dataframe that contains of the 3 dataframe merged together without subsetting them yet
new_merged = pd.merge(ScimEn , pd.merge(energy, GDP ,how= 'left', on='Country'),how='left', on='Country')
new_merged
# checking the new dataframe lenght
len(new_merged)

191

In [30]:
# substracting the length of 'merged_df' from the 'new_merged' df to get the number to get the total number of lost rows
len(new_merged)-len(merged_df)

176

In [31]:
merged_df

Rank  Documents  Citable documents  Citations  \
Country                                                             
China                  1     127050             126767     597237   
United States          2      96661              94747     792274   
Japan                  3      30504              30287     223024   
United Kingdom         4      20944              20357     206091   
Russian Federation     5      18534              18301      34266   
Canada                 6      17899              17620     215003   
Germany                7      17027              16831     140566   
India                  8      15005              14841     128763   
France                 9      13153              12973     130632   
South Korea           10      11983              11923     114675   
Italy                 11      10964              10794     111850   
Spain                 12       9428               9330     123336   
Iran                  13       8896               8819      57470   
Australia             14       8831               8725      90765   
Brazil                15       8668               8596      60702   

                    Self-citations  Citations per document  H index  \
Country                                                               
China                       411683                    4.70      138   
United States               265436                    8.20      230   
Japan                        61554                    7.31      134   
United Kingdom               37874                    9.84      139   
Russian Federation           12422                    1.85       57   
Canada                       40930                   12.01      149   
Germany                      27426                    8.26      126   
India                        37209                    8.58      115   
France                       28601                    9.93      114   
South Korea                  22595                    9.57      104   
Italy                        26661                   10.20      106   
Spain                        23964                   13.08      115   
Iran                         19125                    6.46       72   
Australia                    15606                   10.28      107   
Brazil                       14396                    7.00       86   

                    Energy Supply  Energy Supply per Capita  % Reneweble  \
Country                                                                    
China                1.271910e+11                      93.0    19.754910   
United States                 NaN                       NaN          NaN   
Japan                1.898400e+10                     149.0    10.232820   
United Kingdom                NaN                       NaN          NaN   
Russian Federation   3.070900e+10                     214.0    17.288680   
Canada               1.043100e+10                     296.0    61.945430   
Germany              1.326100e+10                     165.0    17.901530   
India                3.319500e+10                      26.0    14.969080   
France               1.059700e+10                     166.0    17.020280   
South Korea          1.100700e+10                     221.0     2.279353   
Italy                6.530000e+09                     109.0    33.667230   
Spain                4.923000e+09                     106.0    37.968590   
Iran                          NaN                       NaN          NaN   
Australia            5.386000e+09                     231.0    11.810810   
Brazil               1.214900e+10                      59.0    69.648030   

                            2006          2007          2008          2009  \
Country                                                                      
China               3.992331e+12  4.559041e+12  4.997775e+12  5.459247e+12   
United States                NaN           NaN           NaN           NaN   
Japan               5.496542e+12 

-------------------------------------------------------------------------------------------
## STEP 5
### What is the Average `GDP` over the last 10 years for each country?


Getting the average GDP over the last 10 years for each country? (exclude missing values from this calculation.)



In [32]:
# getting the average GDP over the last 10 years per each country
average_gdp = merged_df[['2006', '2007',
                       '2008', '2009', '2010', '2011', '2012', '2013', '2014', '2015']].mean(axis = 1)
average_gdp

Country
China                 6.348609e+12
United States                  NaN
Japan                 5.542208e+12
United Kingdom                 NaN
Russian Federation    1.565459e+12
Canada                1.660647e+12
Germany               3.493025e+12
India                 1.769297e+12
France                2.681725e+12
South Korea           1.106715e+12
Italy                 2.120175e+12
Spain                 1.418078e+12
Iran                           NaN
Australia             1.164043e+12
Brazil                2.189794e+12
dtype: float64

-------------------------------------------------------------------------------------------
## STEP 6
### What is the mean of `Energy Supply per Capita`?
Getting the mean Energy Supply per Capita.


In [33]:
# getting the mean of 'Energy Supply per Capita' column
Energy_supply_per_capita_mean = merged_df['Energy Supply per Capita'].mean()
print(Energy_supply_per_capita_mean)

152.91666666666666


-------------------------------------------------------------------------------------------
## STEP 7
### What country has the maximum `% Reneweble`? And what is the percentage?


Getting the country that has the maximum % Renewable and what is the percentage?



In [34]:
# getting the country with the max value of '% Renewable' column
(merged_df['% Reneweble'].idxmax(), merged_df['% Reneweble'].max())

('Brazil', 69.64803)

In [35]:
# another way of getting the info about the country that has the max % Renewable value

# merged_df.loc[merged_df['% Reneweble'].idxmax()]

-------------------------------------------------------------------------------------------
## STEP 8
### What are the ratios of `Self-Citations` to `Total-Citations`? 

Creating a new column that is the ratio of Self-Citations to Total Citations. 
Then, Getting the maximum value for this new column, and Getting the country that has the highest ratio.


In [64]:
# creating a new column that equals the ratio of Self-Citations to Total Citations and multiply them by 100
merged_df['Ratio'] = (merged_df['Self-citations'] / merged_df['Citations']) * 100
merged_df['Ratio']

Country
China                 68.931262
United States         33.503056
Japan                 27.599720
United Kingdom        18.377319
Russian Federation    36.251678
Canada                19.036944
Germany               19.511119
India                 28.897276
France                21.894329
South Korea           19.703510
Italy                 23.836388
Spain                 19.429850
Iran                  33.278232
Australia             17.193852
Brazil                23.715858
Name: Ratio, dtype: float64

----------------------------------------------------------------------------------------------------------------
### What is the `maximum` value of the `Ratio` column? And which `Country` has the max value?

In [37]:
# getting the max value of the new column 'Ratio' with its country
(merged_df['Ratio'].idxmax() , merged_df['Ratio'].max())

('China', 68.93126179389422)

-------------------------------------------------------------------------------------------
## STEP 9
### What is the distribution of my `% Reneweble` column? (through knowing how many values are above or below the median)

Creating a new column with a 1 if the country's % Renewable value is at or above the median for all countries in the top 15, 
And a 0 if the country's % Renewable value is below the median.


In [38]:
# getting the median value of '% Reneweble' column 
merged_df['% Reneweble'].median()

17.595105

In [65]:
# setting up my conditions
conditions = [
    (merged_df['% Reneweble'] >= merged_df['% Reneweble'].median()),
    (merged_df['% Reneweble'] < merged_df['% Reneweble'].median())
]
# setting up my values
values = ['1' , '0']
# creating a new column
merged_df['High Renew'] = np.select(conditions , values)

---------------------------------------------------------------------
### Checking the distribution of `High Renew` column:

In [40]:
# validating a new column's values
merged_df[['% Reneweble' , 'High Renew']]

% Reneweble High Renew
Country                                   
China                 19.754910          1
United States               NaN          0
Japan                 10.232820          0
United Kingdom              NaN          0
Russian Federation    17.288680          0
Canada                61.945430          1
Germany               17.901530          1
India                 14.969080          0
France                17.020280          0
South Korea            2.279353          0
Italy                 33.667230          1
Spain                 37.968590          1
Iran                        NaN          0
Australia             11.810810          0
Brazil                69.648030          1

----------------------------------------------------------------------------------
## STEP 10 
Using the following dictionary to group the Countries by Continent, 
Then creating a dateframe that displays the sample size (the number of countries in each continent bin), and the sum, mean, and std deviation for the estimated population of each country.


### The dictionary that will be used to group countries by continent:
ContinentDict  = {'China':'Asia', 
                  'United States':'North America', 
                  'Japan':'Asia', 
                  'United Kingdom':'Europe', 
                  'Russian Federation':'Europe', 
                  'Canada':'North America', 
                  'Germany':'Europe', 
                  'India':'Asia',
                  'France':'Europe', 
                  'South Korea':'Asia', 
                  'Italy':'Europe', 
                  'Spain':'Europe', 
                  'Iran':'Asia',
                  'Australia':'Australia', 
                  'Brazil':'South America'}
                  

In [41]:
# defining a new dictionarty 
ContinentDict  = {'China':'Asia', 
                  'United States':'North America', 
                  'Japan':'Asia', 
                  'United Kingdom':'Europe', 
                  'Russian Federation':'Europe', 
                  'Canada':'North America', 
                  'Germany':'Europe', 
                  'India':'Asia',
                  'France':'Europe', 
                  'South Korea':'Asia', 
                  'Italy':'Europe', 
                  'Spain':'Europe', 
                  'Iran':'Asia',
                  'Australia':'Australia', 
                  'Brazil':'South America'}

In [42]:
# creating a new column 'Continent'
merged_df['Continent'] = pd.Series(ContinentDict)

In [46]:
# validating the values of 'Continent' column
merged_df['Continent']

Country
China                          Asia
United States         North America
Japan                          Asia
United Kingdom               Europe
Russian Federation           Europe
Canada                North America
Germany                      Europe
India                          Asia
France                       Europe
South Korea                    Asia
Italy                        Europe
Spain                        Europe
Iran                           Asia
Australia                 Australia
Brazil                South America
Name: Continent, dtype: object

In [66]:
merged_df

Rank  Documents  Citable documents  Citations  \
Country                                                             
China                  1     127050             126767     597237   
United States          2      96661              94747     792274   
Japan                  3      30504              30287     223024   
United Kingdom         4      20944              20357     206091   
Russian Federation     5      18534              18301      34266   
Canada                 6      17899              17620     215003   
Germany                7      17027              16831     140566   
India                  8      15005              14841     128763   
France                 9      13153              12973     130632   
South Korea           10      11983              11923     114675   
Italy                 11      10964              10794     111850   
Spain                 12       9428               9330     123336   
Iran                  13       8896               8819      57470   
Australia             14       8831               8725      90765   
Brazil                15       8668               8596      60702   

                    Self-citations  Citations per document  H index  \
Country                                                               
China                       411683                    4.70      138   
United States               265436                    8.20      230   
Japan                        61554                    7.31      134   
United Kingdom               37874                    9.84      139   
Russian Federation           12422                    1.85       57   
Canada                       40930                   12.01      149   
Germany                      27426                    8.26      126   
India                        37209                    8.58      115   
France                       28601                    9.93      114   
South Korea                  22595                    9.57      104   
Italy                        26661                   10.20      106   
Spain                        23964                   13.08      115   
Iran                         19125                    6.46       72   
Australia                    15606                   10.28      107   
Brazil                       14396                    7.00       86   

                    Energy Supply  Energy Supply per Capita  % Reneweble  ...  \
Country                                                                   ...   
China                1.271910e+11                      93.0    19.754910  ...   
United States                 NaN                       NaN          NaN  ...   
Japan                1.898400e+10                     149.0    10.232820  ...   
United Kingdom                NaN                       NaN          NaN  ...   
Russian Federation   3.070900e+10                     214.0    17.288680  ...   
Canada               1.043100e+10                     296.0    61.945430  ...   
Germany              1.326100e+10                     165.0    17.901530  ...   
India                3.319500e+10                      26.0    14.969080  ...   
France               1.059700e+10                     166.0    17.020280  ...   
South Korea          1.100700e+10                     221.0     2.279353  ...   
Italy                6.530000e+09                     109.0    33.667230  ...   
Spain                4.923000e+09                     106.0    37.968590  ...   
Iran                          NaN                       NaN          NaN  ...   
Australia            5.386000e+09                     231.0    11.810810  ...   
Brazil               1.214900e+10                      59.0    69.648030  ...   

                            2010          2011          2012          2013  \
Country                                                                      
China               6.039659e+12  6.612490e+12  7.124978e+12  7.672448e+12   
United States             

--------------------------------------------------------------------------------------------
### What is the `size`, `Sum`, `mean`and the `STD` for the `estimated population` for each country? 

In [49]:
# creating a new colum 'population' that equals the 'Energy Supply' divided by 'Energy Supply per capita'
merged_df['population']=merged_df['Energy Supply'] / merged_df['Energy Supply per Capita']

In [51]:
# showing the new column
merged_df['population']

Country
China                 1.367645e+09
United States                  NaN
Japan                 1.274094e+08
United Kingdom                 NaN
Russian Federation    1.435000e+08
Canada                3.523986e+07
Germany               8.036970e+07
India                 1.276731e+09
France                6.383735e+07
South Korea           4.980543e+07
Italy                 5.990826e+07
Spain                 4.644340e+07
Iran                           NaN
Australia             2.331602e+07
Brazil                2.059153e+08
Name: population, dtype: float64

In [60]:
# getting the (size,sum,mean,std) for the 'population' column and putting them into a new df
population_df = merged_df.groupby('Continent')['population'].agg([np.size ,sum ,np.mean , np.std])

In [61]:
population_df

size           sum          mean           std
Continent                                                    
Asia              5  2.821591e+09  7.053977e+08  7.138779e+08
Australia         1  2.331602e+07  2.331602e+07           NaN
Europe            6  3.940587e+08  7.881174e+07  3.813228e+07
North America     2  3.523986e+07  3.523986e+07           NaN
South America     1  2.059153e+08  2.059153e+08           NaN